packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts

```diff
@@ -154,32 +157,57 @@ export class OptimisticOracleClient {
 
   public async update(): Promise<void> {
     // Determine earliest block to query events for based on lookback window:
+    if (this.chainId === -1) this.chainId = await this.web3.eth.getChainId();
     const [averageBlockTime, currentBlock] = await Promise.all([
-      averageBlockTimeSeconds(),
+      averageBlockTimeSeconds(this.chainId),
       this.web3.eth.getBlock("latest"),
     ]);
     const lookbackBlocks = Math.ceil(this.lookback / averageBlockTime);
     const earliestBlockToQuery = Math.max(currentBlock.number - lookbackBlocks, 0);
 
-    // Fetch contract state variables in parallel.
-    // Note: We can treat all events by default as normal OptimisticOracle events because in most cases we only
-    // need to read the requester, identifier, timestamp, and ancillary data which are emitted in both Skinny
-    // and normal OptimisticOracle events.
-    const [requestEvents, proposalEvents, disputeEvents, settleEvents, currentTime] = await Promise.all([
-      (this.oracle.getPastEvents("RequestPrice", { fromBlock: earliestBlockToQuery }) as unknown) as Promise<
-        RequestPrice[] | SkinnyRequestPrice[]
-      >,
-      (this.oracle.getPastEvents("ProposePrice", { fromBlock: earliestBlockToQuery }) as unknown) as Promise<
-        ProposePrice[] | SkinnyProposePrice[]
-      >,
-      (this.oracle.getPastEvents("DisputePrice", { fromBlock: earliestBlockToQuery }) as unknown) as Promise<
-        DisputePrice[] | SkinnyDisputePrice[]
-      >,
-      (this.oracle.getPastEvents("Settle", { fromBlock: earliestBlockToQuery }) as unknown) as Promise<
-        Settle[] | SkinnySettle[]
-      >,
-      this.oracle.methods.getCurrentTime().call(),
+    const eventResults = await Promise.all([
+      getEventsWithPaginatedBlockSearch(
```

Comment:
super clean with this util. good work. 

one thing I'm not a huge fan of is the type casting you have to do in all these place with the `(this.oracle as unknown) as Web3Contract`. that's kinda painful. is there a way we dont have to do this? `this.oracle` is 
`export type OptimisticOracleContract = SkinnyOptimisticOracleWeb3 | OptimisticOracleWeb3`. is this not of type `Web3Contract`?

Reply:
Nah its not quite a type of Web3Contract